tmt/policy.py

```diff
@@ -0,0 +1,222 @@
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any, Optional, TypeVar
+
+import tmt.container
+import tmt.utils
+from tmt._compat.pydantic import ValidationError
+from tmt.container import Extra, MetadataContainer, metadata_field
+from tmt.log import Logger, Topic
+from tmt.utils import FieldValueSource, Path, ShellScript
+from tmt.utils.templates import render_template
+
+if TYPE_CHECKING:
+    from tmt.base import Core, Test
+
+T = TypeVar('T')
+
+
+#: A template showing changes made by an instruction.
+KEY_DIFF_TEMPLATE = """
+{{ OLD_VALUE | to_yaml | prefix('- ') | style(fg='red') | trim }}
+{{ NEW_VALUE | to_yaml | prefix('+ ') | style(fg='green') | trim }}
+
+Field value source changed from {{ OLD_VALUE_SOURCE.value | style(fg='red') }} to {{ NEW_VALUE_SOURCE.value | style(fg='green') }}
+"""  # noqa: E501
+
+
+class Instruction(MetadataContainer, extra=Extra.allow):
+    """
+    A single instruction describing changes to test, plan or story keys.
+    """
+
+    def apply(self, obj: 'Core', logger: Logger) -> None:
+        """
+        Apply the instruction to a given object.
+
+        :param obj: object to modify - a test, plan, or story.
+        :param logger: used for logging.
+        """
+
+        base_logger = logger
+
+        def set_key(
+            key: str,
+            template: str,
+            current_value_exported: Any,
+            current_value_source: FieldValueSource,
+            normalize_callback: tmt.container.NormalizeCallback[T],
+        ) -> T:
+            """
+            Update a single key of the object.
+
+            :param key: name of the key to update.
+            :param template: template rendering to a new value.
+            :param current_value_exported: current value as if exported
+                via :py:meth:`Core._export`. Consists of Python built-in
+                types only, no custom classes.
+            :param current_value_source: origin of the current value.
+            :param normalize_callback: will be called with the rendered
+                ``template`` to produce new value for the key.
+            """
+
+            rendered_new_value = render_template(
+                template,
+                VALUE=current_value_exported,
+                VALUE_SOURCE=current_value_source,
```

Comment:
Would be nice to exercise the `VALUE_SOURCE` at least in a simple test. Perhaps something like checking the `5m` as the default `duration` versus explicitly set by user in the config?

Reply:
Added in b39721c0.